internal/apimeta/mapper.go

```diff
@@ -0,0 +1,81 @@
+package apimeta
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+// ResourceMapper allows callers to map resources to kinds.
+type ResourceMapper interface {
+	// NamespacedKindFor returns the priority mapping GVK for a namespaced GR.
+	// If the GR doesn't exist, or isn't namespaced, returns an error.
+	NamespacedKindFor(gr schema.GroupResource) (schema.GroupVersionKind, error)
+}
+
+// NotNamespacedError is returned if the mapper can find the resource, but it is not namespaced.
+type NotNamespacedError struct {
+	GroupResource schema.GroupResource
+}
+
+func (e *NotNamespacedError) Error() string {
+	return fmt.Sprintf("resource %v is not namespaced", e.GroupResource)
+}
+
+func IsNotNamespacedError(err error) bool {
+	if err == nil {
+		return false
+	}
+	switch err.(type) {
+	case *NotNamespacedError:
+		return true
+	default:
+		return false
+	}
+}
+
+func NewAPIMapper(c *rest.Config) (ResourceMapper, error) {
```

Comment:
Bikeshedding time - instead of `APIMapper` (which is very generic), how about `GroupKindMapper`? That's far more descriptive of what it actually does.

You can then call the variables or fields `gkMapper` which might read a little better than simply `mapper`.

Reply:
One of the reasons I created this PR in the first place, was exactly to map more that just GR to (namespaced) GVK. 🙄 And to factor out some low-level technical code from the reconcilers - to make it easier to read/understand. But I can probably try to improve the validation of objects without doing anything with this mapping....

Reply:
Ok, up to you. I'd be happy to change the name to something specific today and then generalize it when it does more (these are all fully internal names, after all) but I don't feel too strongly.

Reply:
Changed name to `GroupKindMapper` - as suggested.